File: week-2-100-doors.hs
doors = [1..100]


countDoorsOpen :: [Int] -> Int
countDoorsOpen doorIndices = (sum . map isOpen) doorIndices


isOpen :: Int -> Int  -- returns 1 for open, 0 for closed
isOpen n = mod nVisits 2  -- door left open if 'visited' an odd number of times
    where nVisits = (sum . mapDivisibilityTest) [1..n]
          mapDivisibilityTest = map (\n2 -> if mod n n2 == 0 then 1 else 0)
          -- an integer n2 'visits' n if and only if n2 divides n


test = 10 == (countDoorsOpen doors)

File: week-2-100-doors.clj
; 100 Doors

(def doors (map (fn [_] :closed) (range 1 101)))

(defn toggle-door [door]
  (if (= door :open) :closed :open))

(defn divisible? [index current-pass]
  (= 0 (mod index current-pass)))

(defn hundred-doors [doors current-pass]
  (if (= 100 current-pass)
    doors
    (recur
      (map-indexed
        (fn [index, door]
          (if (divisible? index current-pass) (toggle-door door) door))
        doors)
      (+ 1 current-pass))))

(hundred-doors doors 1)
